src/base/TypeChecker.ml

```diff
@@ -329,7 +329,12 @@ module ScillaTypechecker (SR : Rep) (ER : Rep) = struct
                (mk_qual_tp t, rep) ),
              remaining_gas )
     | TFun (tvar, body) ->
-        let tenv' = TEnv.addV (TEnv.copy tenv) tvar in
+        let id = get_id tvar in
+        let tenv_del =
+          TEnv.filterTs (TEnv.copy tenv) ~f:(fun _ rr ->
+              String.(pp_typ (rr_typ rr).tp <> id))
```

Comment:
This removes `x` from the type environment, but that isn't the correct fix because of stuff like this:
```
fun ignore_first : forall 'A. forall 'B. 'A -> 'B -> 'B =
  tfun 'A => tfun 'B =>
    fun (first : 'A) -> fun (second : 'B) -> second

let id : forall 'A. 'A -> 'A =
  tfun 'A => fun (x : 'A) =>
    let my_partial_ignore_first = @ignore_first 'A in
    let consistent : forall 'A. 'A -> 'A  =
      tfun 'A =>
        fun (my_second : 'A) ->
          let my_full_ignore_first = @my_partial_ignore_first 'A in
          my_full_ignore_first x my_second (* No longer a  type error *)
    in
    ...
```

Also, it's not a great error message to say that `x` isn't bound in the type environment, when `x` has actually been properly declared. 

The real reason the program is illegal is that if you alpha-rename the type of the TFun, then `x` still retains the type from the surrounding environment. This suggests that the fix is to perform the alpha-renaming of `tvar` inside `body`, rather than removing `x` from the type environment of `body`.

Alpha-renaming is only necessary if `List.find TEnv.tvars ~f:(fun v _ -> v = id)` returns `Some _`, but the actual renaming needs to be handled on the fly (probably inside TEnv) because we can't typecheck the body of the TFun without the existing type environment.

Reply:
Alternatively, if we want a quick fix, we could make it illegal to declare a new type variable inside the scope of another type variable with the same name. Something like

```
  match List.find TEnv.tvars ~f:(fun v _ -> v = id) with
  | Some _ -> (* Error *)
  | None -> (* Update the type environment as before, and typecheck the body *)
```

Note also that none of these changes are strictly speaking backward compatible, but I think this falls under the category of bug fixes, so I think it should be allowed.

Reply:
You are absolutely right, I don't know what I was thinking. Should not have been writing code that late :)
All the proper designs I could come up with require too many changes, some of them including `typ`, which I would like to avoid for now. I just implemented your suggestion about forbidding nested equal type identifiers. Some tests needed a bit of adjustment. Please have another look.